### PR TITLE
Configuration resolver review

### PR DESCRIPTION
**Pull request by SpacePossum**
Hi,

Thanks for all the work on the new configuration options.
I'm opening this messy PR early as a way of discussing some ideas of mine. Please take no offense, I'm just trying to see if we can make your good work even more awesome.
I'll will be adding comments on my own changes so we can discuss those. If you don't want to than that is fine as well of course :)

### CODE REVIEW

File: src/AbstractFixer.php [hunk 1]
```diff
@@ -89,21 +89,13 @@ public function supports(\SplFileInfo $file)
...
-     */
     public function configure(array $configuration = null)
     {
         if (!$this instanceof ConfigurationDefinitionFixerInterface) {
-            throw new \LogicException('Cannot run method for class not implementing `ConfigurationDefinitionFixerInterface`.');
+            throw new \LogicException(sprintf('Cannot configure using Abstact parent, child not implementing "%s".', ConfigurationDefinitionFixerInterface::class));
         }
 
         if (null === $configuration) {
-            @trigger_error(
```
Comment (SpacePossum) [inline src/AbstractFixer.php]: I think this is what @keradus meant in this comment https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/2500#discussion_r100700016
Deprecating allowing `null` as default configuration is not needed in this PR. If we want to deprecate it we can do this through a separate PR and see what people think of it there. Removing it here will leave less to discuss, meaning we can go forward faster ;)

File: src/Fixer/ConfigurationDefinitionFixerInterface.php [hunk 2]
```diff
@@ -15,6 +15,8 @@
 use PhpCsFixer\OptionsResolver;
 
 /**
+ * Maybe better the other way around, i.e. now there is no upgrade path to the new interface without deprecation warnings
```
Comment (SpacePossum) [inline src/Fixer/ConfigurationDefinitionFixerInterface.php]: This is more of concept and I would like to hear what others think.
By deprecating the old way and at the same time deprecating the new way, it leaves no upgrade path without using deprecations. I think we should try avoid that. I think we can remove the annotation here and add later when we have another upgrade path, for example when 3.0 is released or something similar.

File: src/Fixer/Import/OrderedImportsFixer.php [hunk 3]
```diff
@@ -472,7 +472,6 @@ private function sortByAlgorithm($indexes)
                 break;
             default:
                 throw new \LogicException(sprintf('Sort algorithm "%s" is not supported.', $this->configuration['sortAlgorithm']));
-                break;
```
Comment (SpacePossum) [inline src/Fixer/Import/OrderedImportsFixer.php]: just a minor (I won't comment on all minors, let me know if you've a question about one of those)

File: src/FixerDefinition/FixerDefinition.php [hunk 4]
```diff
@@ -41,9 +41,9 @@ public function __construct(
         $riskyDescription = null
     ) {
         if (6 === func_num_args()) {
-            @trigger_error('Arguments #5 and #6 of FixerDefinition::__construct() is deprecated and will be dropped in 3.0, use argument #4 instead.', E_USER_DEPRECATED);
-        } elseif (5 === func_num_args()) {
-            @trigger_error('Argument #5 of FixerDefinition::__construct() is deprecated and will be dropped in 3.0.', E_USER_DEPRECATED);
+            @trigger_error('Arguments #5 and #6 of FixerDefinition::__construct() is deprecated and will be removed in 3.0, use argument #4 instead.', E_USER_DEPRECATED);
+        } elseif (5 === func_num_args()) { // dropped -> removed
```
Comment (SpacePossum) [inline src/FixerDefinition/FixerDefinition.php]: oh I didn't clean up this comment, will remove soon

File: tests/Fixer/Alias/NoMixedEchoPrintFixerTest.php [hunk 5]
```diff
@@ -292,7 +292,7 @@ public function testWrongConfig($wrongConfig, $expectedMessage)
     {
         $this->setExpectedException(
             'PhpCsFixer\ConfigurationException\InvalidFixerConfigurationException',
-            $expectedMessage
+            $expectedMessage // should be reverted, the regex is more explicit and should stay
```
Comment (SpacePossum) [inline tests/Fixer/Alias/NoMixedEchoPrintFixerTest.php]: I reverted a bunch of changes in the test where the `RegEx` was dropped. I feel these have value as these test more explicit. Loosing up the test is not needed so I would like not to do it.

File: tests/Fixer/ClassNotation/SingleClassElementPerStatementFixerTest.php [hunk 6]
```diff
@@ -606,9 +606,7 @@ class Foo
 EOT;
 
         $this->fixer->configure($configuration);
-        $this->doTest($expected, $input);
 
-        $this->fixer->configure(array('elements' => $configuration));
```
Comment (SpacePossum) [inline tests/Fixer/ClassNotation/SingleClassElementPerStatementFixerTest.php]: I think it is better not to test two thinks in one test, therefor I think it is better to move this into a new test method.

File: tests/Fixer/FunctionNotation/ReturnTypeDeclarationFixerTest.php [hunk 7]
```diff
@@ -44,10 +44,6 @@ public function testFixWithDefaultConfiguration($expected, $input = null)
         $this->fixer->configure(null);
 
         $this->doTest($expected, $input);
-
-        $this->fixer->configure(array());
-
-        $this->doTest($expected, $input);
```
Comment (SpacePossum) [inline tests/Fixer/FunctionNotation/ReturnTypeDeclarationFixerTest.php]: Testing if `null` and `array` result in the same thing is testing if the `AbstractFixer` handles this correctly. I think this should be done in a diff. test or split up here into multiple test methods.

File: tests/Fixer/Whitespace/NoExtraConsecutiveBlankLinesFixerTest.php [hunk 8]
```diff
@@ -146,6 +145,12 @@ public function provideConfigTests()
         }
         $tests[] = $all;
 
+        // default configuration test
+        $tests[] = array(
```
Comment (SpacePossum) [inline tests/Fixer/Whitespace/NoExtraConsecutiveBlankLinesFixerTest.php]: I'm not sure why this test case was removed.

File: src/OptionsResolver.php [hunk 9]
```diff
@@ -113,7 +113,7 @@ public function getDefault($option)
      */
     public function getDefaults()
     {
-        return $this->defaults;
+        return $this->defaults; // maybe we get to SF to break open the parent?
     }
```
Comment (SpacePossum) [inline src/OptionsResolver.php]: I noticed that a few properties in this class are the same as the parents.

For the `normalizers` I wonder if this one can be removed add done through the parent completely.

For the `defaults` I see you need it as the parent doesn't allow to get the `defaults`. I wonder if maybe this can be changed in the SF class itself. However that is of course out of scope. I just forgot to remove the comment.

**Comment by SpacePossum** [comment]
For the test it would be nice to test the deprecations, as now there are a lot of warning of which some can be avoided.
This is the list on my setup:
```
Passing control_statements at the root of the configuration is deprecated and will not be supported in 3.0, use "control_statements" => array(...) option instead: 329x
    328x in NoUnneededControlParenthesesFixerTest::testFix from PhpCsFixer\Tests\Fixer\ControlStructure
    1x in FixerFactoryTest::testFixerDefinitions from PhpCsFixer\Tests

Passing assertions at the root of the configuration is deprecated and will not be supported in 3.0, use "assertions" => array(...) option instead: 182x
    150x in PhpUnitConstructFixerTest::testFix from PhpCsFixer\Tests\Fixer\PhpUnit
    29x in PhpUnitStrictFixerTest::testFix from PhpCsFixer\Tests\Fixer\PhpUnit
    1x in PhpUnitConstructFixerTest::testInvalidConfiguration from PhpCsFixer\Tests\Fixer\PhpUnit
    1x in PhpUnitConstructFixerTest::testInvalidConfig from PhpCsFixer\Tests\Fixer\PhpUnit
    1x in FixerFactoryTest::testFixerDefinitions from PhpCsFixer\Tests

PhpCsFixer\FixerDefinition\FixerDefinition::getConfigurationDescription is deprecated and will be removed in 3.0: 135x
    132x in FixerFactoryTest::testFixerDefinitions from PhpCsFixer\Tests
    3x in FixerDefinitionTest::testGetConfigurationDescription from PhpCsFixer\Tests\FixerDefinition

PhpCsFixer\FixerDefinition\FixerDefinition::getDefaultConfiguration is deprecated and will be removed in 3.0: 134x
    132x in FixerFactoryTest::testFixerDefinitions from PhpCsFixer\Tests
    2x in FixerDefinitionTest::testGetDefaultConfiguration from PhpCsFixer\Tests\FixerDefinition

Passing tokens at the root of the configuration is deprecated and will not be supported in 3.0, use "tokens" => array(...) option instead: 59x
    11x in FixerFactoryTest::testFixerDefinitions from PhpCsFixer\Tests
    10x in IntegrationTest::testIntegration from PhpCsFixer\Tests
    8x in NoExtraConsecutiveBlankLinesFixerTest::testWithConfig from PhpCsFixer\Tests\Fixer\Whitespace
    7x in NoExtraConsecutiveBlankLinesFixerTest::testBetweenUse from PhpCsFixer\Tests\Fixer\Whitespace
    6x in NoExtraConsecutiveBlankLinesFixerTest::testBraces from PhpCsFixer\Tests\Fixer\Whitespace
    4x in NoExtraConsecutiveBlankLinesFixerTest::testWithoutUses from PhpCsFixer\Tests\Fixer\Whitespace
    4x in NoExtraConsecutiveBlankLinesFixerTest::testOneOrInLineCases from PhpCsFixer\Tests\Fixer\Whitespace
    2x in NoExtraConsecutiveBlankLinesFixerTest::testMessyWhitespaces from PhpCsFixer\Tests\Fixer\Whitespace
    1x in NoExtraConsecutiveBlankLinesFixerTest::testWrongConfig from PhpCsFixer\Tests\Fixer\Whitespace
    1x in NoExtraConsecutiveBlankLinesFixerTest::testRemoveLinesBetweenUseStatements from PhpCsFixer\Tests\Fixer\Whitespace
    1x in NoExtraConsecutiveBlankLinesFixerTest::testRemoveLinesBetweenUseStatements70 from PhpCsFixer\Tests\Fixer\Whitespace
    1x in NoExtraConsecutiveBlankLinesFixerTest::testRemoveBetweenUseTraits from PhpCsFixer\Tests\Fixer\Whitespace
    1x in NoExtraConsecutiveBlankLinesFixerTest::testOneOrInLine70Cases from PhpCsFixer\Tests\Fixer\Whitespace
    1x in NoExtraConsecutiveBlankLinesFixerTest::testBraces54 from PhpCsFixer\Tests\Fixer\Whitespace
    1x in RuleSetTest::testRiskyRulesInSet from PhpCsFixer\Tests

Passing replacements at the root of the configuration is deprecated and will not be supported in 3.0, use "replacements" => array(...) option instead: 40x
    6x in PhpdocReturnSelfReferenceFixerTest::testGeneratedFix from PhpCsFixer\Tests\Fixer\Phpdoc
    5x in RandomApiMigrationFixerTest::testFix from PhpCsFixer\Tests\Fixer\Alias
    4x in PhpdocNoAliasTagFixerTest::testTypeToVarFix from PhpCsFixer\Tests\Fixer\Phpdoc
    4x in PhpdocNoAliasTagFixerTest::testVarToTypeFix from PhpCsFixer\Tests\Fixer\Phpdoc
    4x in IntegrationTest::testIntegration from PhpCsFixer\Tests
    3x in PhpdocNoAliasTagFixerTest::testPropertyFix from PhpCsFixer\Tests\Fixer\Phpdoc
    2x in PhpdocReturnSelfReferenceFixerTest::testInvalidConfiguration from PhpCsFixer\Tests\Fixer\Phpdoc
    2x in FixerFactoryTest::testFixerDefinitions from PhpCsFixer\Tests
    1x in PhpdocNoAliasTagFixerTest::testInvalidConfigCase2 from PhpCsFixer\Tests\Fixer\Phpdoc
    1x in PhpdocNoAliasTagFixerTest::testInvalidConfigCase3 from PhpCsFixer\Tests\Fixer\Phpdoc
    1x in PhpdocNoAliasTagFixerTest::testInvalidConfigCase4 from PhpCsFixer\Tests\Fixer\Phpdoc
    1x in RandomApiMigrationFixerTest::testConfigureCheckSearchFunction from PhpCsFixer\Tests\Fixer\Alias
    1x in PhpdocNoAliasTagFixerTest::testInvalidConfigCase1 from PhpCsFixer\Tests\Fixer\Phpdoc
    1x in RandomApiMigrationFixerTest::testConfigure from PhpCsFixer\Tests\Fixer\Alias
    1x in PhpdocNoAliasTagFixerTest::testLinkToSee from PhpCsFixer\Tests\Fixer\Phpdoc
    1x in PhpdocReturnSelfReferenceFixerTest::testFix from PhpCsFixer\Tests\Fixer\Phpdoc
    1x in RandomApiMigrationFixerTest::testConfigureCheckReplacementType from PhpCsFixer\Tests\Fixer\Alias
    1x in PhpdocNoAliasTagFixerTest::testInvalidConfigCase5 from PhpCsFixer\Tests\Fixer\Phpdoc

Passing functions at the root of the configuration is deprecated and will not be supported in 3.0, use "functions" => array(...) option instead: 37x
    32x in PhpUnitDedicateAssertFixerTest::testInternalTypeMethods from PhpCsFixer\Tests\Fixer\PhpUnit
    2x in IntegrationTest::testIntegration from PhpCsFixer\Tests
    1x in PhpUnitDedicateAssertFixerTest::testConfig from PhpCsFixer\Tests\Fixer\PhpUnit
    1x in PhpUnitDedicateAssertFixerTest::testInvalidConfig from PhpCsFixer\Tests\Fixer\PhpUnit
    1x in FixerFactoryTest::testFixerDefinitions from PhpCsFixer\Tests

Passing elements at the root of the configuration is deprecated and will not be supported in 3.0, use "elements" => array(...) option instead: 17x
    4x in VisibilityRequiredFixerTest::testFixClassConst from PhpCsFixer\Tests\Fixer\ClassNotation
    4x in IntegrationTest::testIntegration from PhpCsFixer\Tests
    3x in SingleClassElementPerStatementFixerTest::testFixWithConfiguration from PhpCsFixer\Tests\Fixer\ClassNotation
    2x in FixerFactoryTest::testFixerDefinitions from PhpCsFixer\Tests
    1x in SingleClassElementPerStatementFixerTest::testWrongConfig from PhpCsFixer\Tests\Fixer\ClassNotation
    1x in VisibilityRequiredFixerTest::testInvalidConfigurationType from PhpCsFixer\Tests\Fixer\ClassNotation
    1x in VisibilityRequiredFixerTest::testInvalidConfigurationValue from PhpCsFixer\Tests\Fixer\ClassNotation
    1x in RuleSetTest::testRiskyRulesInSet from PhpCsFixer\Tests

Passing positions at the root of the configuration is deprecated and will not be supported in 3.0, use "positions" => array(...) option instead: 6x
    3x in NoSpacesAroundOffsetFixerTest::testFixWithConfiguration from PhpCsFixer\Tests\Fixer\Whitespace
    2x in FixerFactoryTest::testFixerDefinitions from PhpCsFixer\Tests
    1x in NoSpacesAroundOffsetFixerTest::testWrongConfig from PhpCsFixer\Tests\Fixer\Whitespace

Passing order at the root of the configuration is deprecated and will not be supported in 3.0, use "order" => array(...) option instead: 4x
    3x in OrderedClassElementsFixerTest::testFixWithConfiguration from PhpCsFixer\Tests\Fixer\ClassNotation
    1x in OrderedClassElementsFixerTest::testWrongConfig from PhpCsFixer\Tests\Fixer\ClassNotation

Argument #5 of FixerDefinition::__construct() is deprecated and will be removed in 3.0: 2x
    1x in FixerDefinitionTest::testGetConfigurationDescription from PhpCsFixer\Tests\FixerDefinition
    1x in FixerDefinitionTest::testGetDefaultConfiguration from PhpCsFixer\Tests\FixerDefinition

Arguments #5 and #6 of FixerDefinition::__construct() is deprecated and will be removed in 3.0, use argument #4 instead: 1x
    1x in FixerDefinitionTest::testGetRiskyDescription from PhpCsFixer\Tests\FixerDefinition

Passing annotations at the root of the configuration is deprecated and will not be supported in 3.0, use "annotations" => array(...) option instead: 1x
    1x in FixerFactoryTest::testFixerDefinitions from PhpCsFixer\Tests

Passing foo at the root of the configuration is deprecated and will not be supported in 3.0, use "foo" => array(...) option instead: 1x
    1x in OptionsResolverTest::testMapRootConfigurationTo from PhpCsFixer\Tests

```

Secondly, my PR breaks the unit test of `PhpdocNoAliasTagFixerTest`, however maybe the test is right and something doesn't work anymore as it used to. Would be nice to look more into this.

File: src/AbstractFixer.php (hunk 1 above)
Comment (keradus) [reply src/AbstractFixer.php]: OK, so.
Actually, I thought that we could split one method of master branch - `configure`, which sometimes does the configuration and sometimes removes it, into 2 methods - `configure` and `reset`.
I thought that we would not run `resolver` when one want to reset the configuration.

But since we keep the default config in resolver we would call it in `reset` method anyway. For that, indeed it will be the same as `configure` method, so no need for 2nd method indeed.

then, maybe indeed passing `null` should be deprecated

**Comment by keradus** [comment]
@julienfalque, are you using gitter or sth ?

**Comment by julienfalque** [comment]
> Please take no offense

On the contrary, I'm glad we can team work on this :)

@keradus Nope, but I'm gonna give it a try.

File: src/AbstractFixer.php (hunk 1 above)
Comment (julienfalque) [reply src/AbstractFixer.php]: Yes I keep thinking that passing `null` should be deprecated. But I have no preference whether we should do it in a separate PR. @keradus, your call :)

File: src/AbstractFixer.php [hunk 10]
```diff
@@ -89,21 +89,13 @@ public function supports(\SplFileInfo $file)
...
     }
 
-    /**
-     * {@inheritdoc}
-     */
     public function configure(array $configuration = null)
     {
         if (!$this instanceof ConfigurationDefinitionFixerInterface) {
-            throw new \LogicException('Cannot run method for class not implementing `ConfigurationDefinitionFixerInterface`.');
+            throw new \LogicException(sprintf('Cannot configure using Abstact parent, child not implementing "%s".', ConfigurationDefinitionFixerInterface::class));
```
Comment (julienfalque) [inline src/AbstractFixer.php]: I think this new message is a bit confused, but I agree the previous one was worse. By the way, using `::class` won't work on PHP lower than 5.5.

File: src/Fixer/ConfigurationDefinitionFixerInterface.php (hunk 2 above)
Comment (julienfalque) [reply src/Fixer/ConfigurationDefinitionFixerInterface.php]: Agreed, but if we deprecate this in 3.0, that means that integrating these changes won't be actually complete until 4.0, right? Isn't there another solution so that all deprecations can be removed in 3.0 without adding others?

File: src/AbstractFixer.php (hunk 10 above)
Comment (SpacePossum) [reply src/AbstractFixer.php]: ah yes, keep forgetting about 5.5, I'll revert this one :+1:

File: src/FixerDefinition/FixerDefinition.php [hunk 11]
```diff
@@ -41,9 +41,9 @@ public function __construct(
         $riskyDescription = null
     ) {
         if (6 === func_num_args()) {
-            @trigger_error('Arguments #5 and #6 of FixerDefinition::__construct() is deprecated and will be dropped in 3.0, use argument #4 instead.', E_USER_DEPRECATED);
-        } elseif (5 === func_num_args()) {
-            @trigger_error('Argument #5 of FixerDefinition::__construct() is deprecated and will be dropped in 3.0.', E_USER_DEPRECATED);
+            @trigger_error('Arguments #5 and #6 of FixerDefinition::__construct() is deprecated and will be removed in 3.0, use argument #4 instead.', E_USER_DEPRECATED);
```
Comment (julienfalque) [inline src/FixerDefinition/FixerDefinition.php]:
```diff
-is
+are
```
(my mistake)

File: src/OptionsResolver.php (hunk 9 above)
Comment (julienfalque) [reply src/OptionsResolver.php]: I think they are required because the properties from the parent are private. I don't really like it but I didn't find a better solution for now.

> I wonder if maybe this can be changed in the SF class itself

I do plan to open feature/pull requests on symfony/options-resolver

File: tests/Fixer/ArrayNotation/ArraySyntaxFixerTest.php [hunk 12]
```diff
@@ -25,9 +25,9 @@
 {
     public function testInvalidConfiguration()
     {
-        $this->setExpectedException(
+        $this->setExpectedExceptionRegExp(
             'PhpCsFixer\ConfigurationException\InvalidFixerConfigurationException',
-            '[array_syntax] Invalid configuration: The option "a" does not exist.'
+            '#^\[array_syntax\] Invalid configuration: The option \"a\" does not exist\. Defined options are: \"syntax\"\.$#'
```
Comment (julienfalque) [inline tests/Fixer/ArrayNotation/ArraySyntaxFixerTest.php]: The error message is slightly different when using `symfony/options-revolver` 2.6: the word _Defined_ is replaced with another one I don't rembember. Same for next tests.

File: src/Fixer/ConfigurationDefinitionFixerInterface.php (hunk 2 above)
Comment (SpacePossum) [reply src/Fixer/ConfigurationDefinitionFixerInterface.php]: I would deprecate it here but in 2.x as soon as 3.0 is released (so no deprecating in 3.0)
(if that makes sense..)

File: tests/Fixer/Phpdoc/PhpdocNoAliasTagFixerTest.php [hunk 13]
```diff
@@ -25,39 +25,39 @@
...
         $this->fixer->configure(array('see' => 'link*/'));
     }
 
     public function testInvalidConfigCase4()
     {
-        $this->setExpectedException(
+        $this->setExpectedExceptionRegExp(
             'PhpCsFixer\ConfigurationException\InvalidFixerConfigurationException',
-            '[phpdoc_no_alias_tag] Invalid configuration: Cannot change tag "see" to tag "link", as the tag is set configured to be replaced to "see".'
+            '#^\[phpdoc_no_alias_tag\] Invalid configuration: Cannot change tag \"link\" to tag "see", as the tag is set configured to be replaced to \"link\"\.$#'
```
Comment (julienfalque) [inline tests/Fixer/Phpdoc/PhpdocNoAliasTagFixerTest.php]: Why did you swap _link_ with _see_?

File: tests/Fixer/Whitespace/NoExtraConsecutiveBlankLinesFixerTest.php [hunk 14]
```diff
@@ -97,11 +97,10 @@ private function test123(){
 EOF;
         $this->doTest($this->removeLinesFromString($template, $lineNumberRemoved), $template);
 
-        if (null === $config) {
-            $this->fixer->configure(array());
-        } else {
-            $this->fixer->configure(array('tokens' => $config));
-        }
+        null === $config
```
Comment (julienfalque) [inline tests/Fixer/Whitespace/NoExtraConsecutiveBlankLinesFixerTest.php]: Using a ternary expression to run code without using the returned value looks odd to me, can we use

```php
$this->fixer->configure(null === $config ? array() : array('tokens' => $config));
```
instead?

Comment (SpacePossum) [reply tests/Fixer/Whitespace/NoExtraConsecutiveBlankLinesFixerTest.php]: sure man, any construct is fine, guess one of my custom fixers did this

**Comment by julienfalque** [comment]
> For the test it would be nice to test the deprecations

I though about that and I think it would be a good thing too. What about using `symfony/phpunit-bridge`? It allows to write _legacy_ tests that expect deprecation notices.

File: src/AbstractFixer.php (hunk 1 above)
Comment (keradus) [reply src/AbstractFixer.php]: Please do it.

File: tests/Fixer/Whitespace/NoExtraConsecutiveBlankLinesFixerTest.php (hunk 14 above)
Comment (keradus) [reply tests/Fixer/Whitespace/NoExtraConsecutiveBlankLinesFixerTest.php]: big :+1: for @julienfalque 
currently, code leaks a side effect

**Comment by keradus** [comment]
If you want, go ahead.
For me if we will deprecate sth we should stop using it, then no deprecations are rised in utests - it's enough for me, but as said, you could add dedicated tests for deprecations if you want to.

File: tests/Fixer/Phpdoc/PhpdocNoAliasTagFixerTest.php (hunk 13 above)
Comment (SpacePossum) [reply tests/Fixer/Phpdoc/PhpdocNoAliasTagFixerTest.php]: I swapped those back (https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/2500/files#diff-624a87e241e3a10f7227a5d31661dfaeR60) not sure why these needed swapping in the first place

this one https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/2500/files#diff-624a87e241e3a10f7227a5d31661dfaeR74 is also kind of odd, "b" and "see" are configured but the new exception message is set to "see" and "link"

File: src/Fixer/ConfigurationDefinitionFixerInterface.php (hunk 2 above)
Comment (julienfalque) [reply src/Fixer/ConfigurationDefinitionFixerInterface.php]: It would not solve the problem: no upgrade path without deprecations, even if the deprecations are only there for a weak or so.

File: tests/Fixer/Phpdoc/PhpdocNoAliasTagFixerTest.php (hunk 13 above)
Comment (julienfalque) [reply tests/Fixer/Phpdoc/PhpdocNoAliasTagFixerTest.php]: Ah you're right, I didn't remember. This may be a mistake, I'll check.

**Comment by julienfalque** [comment]
@keradus Yes, we should stop using the code we deprecate, but as this tool can also be used as a library, I think testing deprecations makes sense. I'll try to write dedicated tests.

File: src/Fixer/ConfigurationDefinitionFixerInterface.php (hunk 2 above)
Comment (SpacePossum) [reply src/Fixer/ConfigurationDefinitionFixerInterface.php]: when 3.0 is released without deprecations there will an upgrade path right?

Comment (julienfalque) [reply src/Fixer/ConfigurationDefinitionFixerInterface.php]: Yes but only by upgrading to a new major version, are we ok with that?

Edit: as introducing deprecations requires a minor version increment, people won't have unexpected deprecations when upgrading to a new patch version. And when upgrading to said new minor version, they are supposed to know that and check changelogs, so 👍 for your proposal.

File: tests/Fixer/ArrayNotation/ArraySyntaxFixerTest.php (hunk 12 above)
Comment (SpacePossum) [reply tests/Fixer/ArrayNotation/ArraySyntaxFixerTest.php]: please make the smallest catch all group for diff. between the messages

File: src/OptionsResolver.php (hunk 9 above)
Comment (SpacePossum) [reply src/OptionsResolver.php]: great to hear :)
good luck over at SF, let me know if I can help out :)

Final thingy, please make this class `final`, even though marked as `@internal` (:+1:) it is still nice to be explicit.

File: tests/Fixer/ArrayNotation/ArraySyntaxFixerTest.php (hunk 12 above)
Comment (julienfalque) [reply tests/Fixer/ArrayNotation/ArraySyntaxFixerTest.php]: Side question: why do you escape `"` character? As far as I know, it is not a meta-character in PCRE patterns.

Comment (SpacePossum) [reply tests/Fixer/ArrayNotation/ArraySyntaxFixerTest.php]: Indeed, this makes no sense from me

**Comment by julienfalque** [comment]
@SpacePossum I cherry-picked your commit (with changes applied to it). You will need to rebase your branch if you want to continue using it for reviewing. Thanks!

File: src/FixerDefinition/FixerDefinition.php (hunk 4 above)
Comment (julienfalque) [reply src/FixerDefinition/FixerDefinition.php]: I removed it while cherry-picking.

File: tests/Fixer/Alias/NoMixedEchoPrintFixerTest.php (hunk 5 above)
Comment (julienfalque) [reply tests/Fixer/Alias/NoMixedEchoPrintFixerTest.php]: I restored the regexps.

File: src/AbstractFixer.php (hunk 10 above)
Comment (julienfalque) [reply src/AbstractFixer.php]: Fixed the `::class` usage while cherry-picking. I left the exception message unchanged.

File: src/FixerDefinition/FixerDefinition.php (hunk 11 above)
Comment (julienfalque) [reply src/FixerDefinition/FixerDefinition.php]: Fixed while cherry-picking.

File: tests/Fixer/ArrayNotation/ArraySyntaxFixerTest.php (hunk 12 above)
Comment (julienfalque) [reply tests/Fixer/ArrayNotation/ArraySyntaxFixerTest.php]: Fixed while cherry-picking.

File: tests/Fixer/Phpdoc/PhpdocNoAliasTagFixerTest.php (hunk 13 above)
Comment (julienfalque) [reply tests/Fixer/Phpdoc/PhpdocNoAliasTagFixerTest.php]: My changes introduced a subtle difference in the way circular replacements are detected, I reverted it.

Comment (julienfalque) [reply tests/Fixer/Phpdoc/PhpdocNoAliasTagFixerTest.php]: See FriendsOfPHP/PHP-CS-Fixer#2560

File: tests/Fixer/Whitespace/NoExtraConsecutiveBlankLinesFixerTest.php (hunk 14 above)
Comment (SpacePossum) [reply tests/Fixer/Whitespace/NoExtraConsecutiveBlankLinesFixerTest.php]: happy with revert, all good, out of curiosity what is this leaking side effect? (just trying to understand)

Comment (julienfalque) [reply tests/Fixer/Whitespace/NoExtraConsecutiveBlankLinesFixerTest.php]: I think what @keradus meant is that the initial goal of such ternary expression is to choose between two values (no side effect, we compute something and returns it), while here we choose between two actions.

IMO regular `if/else` is more suitable for choosing between actions, at least for readability.

File: tests/Fixer/ClassNotation/SingleClassElementPerStatementFixerTest.php (hunk 6 above)
Comment (julienfalque) [reply tests/Fixer/ClassNotation/SingleClassElementPerStatementFixerTest.php]: "Legacy" tests that test deprecated code are now separate tests.

File: tests/Fixer/Whitespace/NoExtraConsecutiveBlankLinesFixerTest.php (hunk 8 above)
Comment (julienfalque) [reply tests/Fixer/Whitespace/NoExtraConsecutiveBlankLinesFixerTest.php]: Moved to a legacy test.
